src/main/java/org/opensearch/security/filter/SecurityRequest.java

```diff
@@ -49,4 +51,12 @@ default String header(final String headerName) {
 
     /** The parameters associated with this request */
     Map<String, String> params();
+
+    /** Gets the parameter for the given key */
+    String param(String key);
+
+    /** Gets the parameters that have been consumed */
+    default Set<String> getConsumedParams() {
+        return Collections.emptySet();
+    };
```

Comment:
This leaks the abstraction out to the caller.  Lets try to keep the interface as clean as possible.  I think you could return a Map when params is called that overloads the `get()` behavior to consume the parameter before retrieving the value from the map.  What do you think?

Reply:
I need a way to get the parameters have been consumed when the underlying request is a NettyRequest (which it is in the headerverifier). These consumed parameters then get added as a channel attribute so that the SecurityRestFilter can consume them when the underlying request is a RestRequest. One alternative I can see to this implementation would be to modify `params()` to be `Map<ParamKey, ParamValue> params();` and the class for ParamKey could keep track if the key has been consumed or not, but I think this approach is cleaner. Are you thinking of a specific alternative to what's in this PR? I'd love to keep this interface minimal as well.

Reply:
Lets keep iterating on this; let me know how I can help.

Reply:
Looking into alternatives now

Reply:
@peternied I removed this from the interface and left it on the NettyRequest object. In the headerverifier the request is cast to a NettyRequest where I can check to see if any parameters were consumer on this request object and add them to the channel attributes to later be consumed again downstream in the SecurityRestFilter.

Reply:
See https://github.com/opensearch-project/security/pull/3975/files#r1473512100

Reply:
- @cwperks Take a look at https://github.com/cwperks/security/pull/22

Let me know what you think of this approach

Reply:
@peternied Thank you for showing the suggested change. It looks good to me, but I think some of the naming of the methods need to be updated to reflect what the method does. 

This PR has an inline anonymous map [here](https://github.com/opensearch-project/security/pull/3975/files#diff-a0353f22ad00aaeca928ebc7caf2849de38478412c7d19fc860c1d486b9b7c98R73-R79), but I think what you should with the CheckedAccessMap is a nicer way to implement that. You can push any changes to this branch that you think would be good changes.